File: scala/scala-impl/test/org/jetbrains/plugins/scala/base/ScalaLightPlatformCodeInsightTestCaseAdapter.java
package org.jetbrains.plugins.scala.base;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.application.WriteAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.projectRoots.JavaSdkVersion;
import com.intellij.openapi.projectRoots.ProjectJdkTable;
import com.intellij.openapi.projectRoots.Sdk;
import com.intellij.openapi.util.registry.Registry;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import com.intellij.testFramework.LightPlatformCodeInsightTestCase;
import com.intellij.testFramework.LightProjectDescriptor;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.plugins.scala.base.DelegatingProjectDescriptor;
import org.jetbrains.plugins.scala.base.libraryLoaders.*;
import org.jetbrains.plugins.scala.base.ScalaSdkOwner;
import org.jetbrains.plugins.scala.debugger.ScalaVersion;
import org.jetbrains.plugins.scala.debugger.Scala_2_10$;
import org.jetbrains.plugins.scala.util.TestUtils;
import scala.collection.Seq;
import scala.collection.immutable.Vector$;
import scala.collection.mutable.Buffer;

import java.io.IOException;
import java.util.ArrayList;

/**
 * @author Alexander Podkhalyuzin
 * @deprecated use {@link org.jetbrains.plugins.scala.base.ScalaLightCodeInsightFixtureTestAdapter} instead
 */
@Deprecated
public abstract class ScalaLightPlatformCodeInsightTestCaseAdapter extends LightPlatformCodeInsightTestCase implements ScalaSdkOwner {

    private static final ThirdPartyLibraryLoader[] EMPTY_LOADERS_ARRAY = new ThirdPartyLibraryLoader[0];

    protected String sourceRootPath() {
        return null;
    }

    protected final String baseRootPath() {
        return TestUtils.getTestDataPath() + "/";
    }

    protected VirtualFile getSourceRootAdapter() {
        return getSourceRoot();
    }

    @Override
    protected Sdk getProjectJDK() {
        return SmartJDKLoader$.MODULE$.getOrCreateJDK(JavaSdkVersion.JDK_1_8);
    }

    @Override
    public ScalaVersion version() {
        return Scala_2_10$.MODULE$;
    }

    @Override
    public Seq<LibraryLoader> librariesLoaders() {
        ArrayList<LibraryLoader> back = new ArrayList<>();

        ScalaSDKLoader scalaLoader = new ScalaSDKLoader(isIncludeReflectLibrary());
        back.add(scalaLoader);

        String path = sourceRootPath();
        if (path != null) {
            back.add(new SourcesLoader(path));
        }

        Buffer<LibraryLoader> result = scala.collection.JavaConverters.asScalaBuffer(back);
        Seq addLibs = additionalLibraries();
        //noinspection unchecked (because variance)
        result.$plus$plus$eq(addLibs);

        return result;
    }

    @NotNull
    @Override
    protected LightProjectDescriptor getProjectDescriptor() {
        return new ScalaLightProjectDescriptor(){
            @Override
            public void tuneModule(Module module) {
                afterSetUpProject(module);
            }

            @Nullable
            @Override
            public Sdk getSdk() {
                return SmartJDKLoader$.MODULE$.getOrCreateJDK(JavaSdkVersion.JDK_1_8);
            }
        };
    }

    protected void afterSetUpProject(Module module) {
        Registry.get("ast.loading.filter").setValue(true, getTestRootDisposable());
        setUpLibraries(module);
    }


    @Override
    protected void setUp() throws Exception {
        super.setUp();

        TestUtils.disableTimerThread();
    }

    protected boolean isIncludeReflectLibrary() {
        return false;
    }

    protected Seq<LibraryLoader> additionalLibraries() {
        return Vector$.MODULE$.empty();
    }

    protected VirtualFile getVFileAdapter() {
        return getVFile();
    }

    protected Editor getEditorAdapter() {
        return getEditor();
    }

    protected Project getProjectAdapter() {
        return getProject();
    }

    protected Module getModuleAdapter() {
        return getModule();
    }

    protected PsiFile getFileAdapter() {
        return getFile();
    }

    protected PsiManager getPsiManagerAdapter() {
        return getPsiManager();
    }

    protected DataContext getCurrentEditorDataContextAdapter() {
        return getCurrentEditorDataContext();
    }

    protected void executeActionAdapter(String actionId) {
        executeAction(actionId);
    }

    protected void configureFromFileTextAdapter(@NonNls final String fileName,
                                                @NonNls final String fileText) throws IOException {
        configureFromFileText(fileName, StringUtil.convertLineSeparators(fileText));
    }

    @Override
    protected void tearDown() throws Exception {
        try {
            disposeLibraries(getModule());
            Sdk[] allJdks = ProjectJdkTable.getInstance().getAllJdks();
            WriteAction.run(() -> {
                for (Sdk jdk : allJdks) {
                    ProjectJdkTable.getInstance().removeJdk(jdk);
                }
            });
        } finally {
            super.tearDown();
        }
    }
}

File: scala/scala-impl/test/org/jetbrains/plugins/scala/base/ScalaLightPlatformCodeInsightTestCaseAdapter.scala
package org.jetbrains.plugins.scala.base

import com.intellij.openapi.editor.Document
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.projectRoots.{ProjectJdkTable, Sdk}
import com.intellij.openapi.util.registry.Registry
import com.intellij.openapi.util.text.StringUtil
import com.intellij.testFramework.{LightPlatformCodeInsightTestCase, LightProjectDescriptor}
import org.jetbrains.plugins.scala.base.libraryLoaders._
import org.jetbrains.plugins.scala.extensions.inWriteAction
import org.jetbrains.plugins.scala.util.TestUtils

@deprecated
abstract class ScalaLightPlatformCodeInsightTestCaseAdapter
  extends LightPlatformCodeInsightTestCase
    with ScalaSdkOwner {

  protected def sourceRootPath: String = null

  final protected def baseRootPath: String = TestUtils.getTestDataPath + "/"

  override protected def getProjectJDK: Sdk = SmartJDKLoader.getOrCreateJDK()

  override protected def librariesLoaders: Seq[LibraryLoader] = {
    val builder = Seq.newBuilder[LibraryLoader]
    val scalaLoader = new ScalaSDKLoader(isIncludeReflectLibrary)
    builder += scalaLoader
    builder ++= Option(sourceRootPath).map(SourcesLoader)
    builder ++= additionalLibraries
    builder.result()
  }

  protected def isIncludeReflectLibrary = false

  protected def additionalLibraries: Seq[LibraryLoader] = Vector.empty

  protected def sharedProjectToken: SharedTestProjectToken = SharedTestProjectToken.DoNotShare

  // TODO: can we reuse the project between test cases in an isolated class in ScalaLightPlatformCodeInsightTestCaseAdapter inheritors?
  override protected def getProjectDescriptor: LightProjectDescriptor = new ScalaLightProjectDescriptor(sharedProjectToken) {
    override def tuneModule(module: Module): Unit = afterSetUpProject(module)

    override def getSdk: Sdk = ScalaLightPlatformCodeInsightTestCaseAdapter.this.getProjectJDK
  }

  protected def afterSetUpProject(module: Module): Unit = {
    Registry.get("ast.loading.filter").setValue(true, getTestRootDisposable)
    setUpLibraries(module)
  }

  override protected def setUp(): Unit = {
    super.setUp()
    TestUtils.disableTimerThread()
  }

  @throws(classOf[Exception])
  override protected def tearDown(): Unit = try {
    disposeLibraries(getModule)
    inWriteAction {
      ProjectJdkTable.getInstance().removeJdk(getProjectJDK)
    }
  } finally {
    super.tearDown()
  }

  override protected def configureFromFileText(fileName: String, fileText: String): Document = {
    val fileTextNormalized = StringUtil.convertLineSeparators(fileText)
    super.configureFromFileText(fileName, fileTextNormalized)
  }

  //overriding to make java protected method public
  //to be able to mix into other classes (see https://github.com/scala/bug/issues/3564)
  override def getProject: Project = super.getProject
}